Sprint 12 ver. 1

DmitryBrosalin opened the pull request:
- Добавил в приложение работу с базой данных.
- Добавил mpa-рейтинг и жанры для фильмов.

DmitryBrosalin committed:
Sprint 12 ver. 1
- Добавил в приложение работу с базой данных.
- Добавил mpa-рейтинг и жанры для фильмов.

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -0,0 +1,136 @@
+
+    public Film save(Film film) {
+        if ((film.getMpa() != null) && (film.getMpa().getId() < 1 || film.getMpa().getId() > 5)) {
+            throw new BadRequestException("id MPA-рейтинга должен быть от 1 до 5.");
```
А разве есть ограничение на id? Этот же id просто генерируется в базе данных при вставке рейтинга

DmitryBrosalin replied:
Насколько я понимаю, эта информация приходит к нам с запросом на сохранения фильма. Да, в тестах от Яндекса этот параметр всегда указан верно, но если рассматривать общий подход к приложению - то кто мешает пользователю неправильно указать id рейтинга? И если он укажет некорректный id, то эта проверка позволяет избежать добавления фильма с некорректным рейтингом.

slavrikk replied:
Можно просто проверить, есть ли такой фильм в бд по запрашиваемому id или нет, если нет, то кидать ошибку.
Текущая реализация дублирует информацию о данных, источник должен быть только база данных

DmitryBrosalin replied:
Понял логику, спасибо 🙏

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -0,0 +1,136 @@
+        }
+        for (Genre genre: film.getGenres()) {
+            if ((genre != null) && (genre.getId() < 1 || genre.getId() > 6)) {
+                throw new BadRequestException("id жанра должен быть от 1 до 6.");
```
Аналогичный комментарий,  по идее мы сколько угодно можем рейтингов добавить в бд

DmitryBrosalin replied:
Насколько я понял ТЗ, задание не предусматривает функционал добавления новых жанров: используем только те, которые изначально вносим в базу данных (отталкиваясь от тестов в постаме это): 
(1, 'Комедия'),
(2, 'Драма'),
(3, 'Мультфильм'),
(4, 'Триллер'),
(5, 'Документальный'),
(6, 'Боевик')

Получается, если пользователь при добавлении фильма указывает иной id жанра, то возникает некорректный запрос на добавление фильма. И если с рейтингами ситуация жестко фиксированная (всего рейтингов 5 и больше их стать не может), то жанров и поджанров действительно можно придумать бесконечное количество. С такой точки зрения было бы логично добавить функционал добавления жанров, однако задание и тесты, которые пропускают проект на проверку, этого решения не предусматривают (тесты явно рассчитаны только на 6 жанров). Поэтому мне кажется в тех условиях, в которых вы находимся, было бы правильнее оставить эту проверку.

slavrikk replied:
Аналогичный комментарий - искать жанр в базе данных, если его нет, кидать ошибку

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    public Collection<Film> getPopularFilms(long size) {
+        return findAll().stream().sorted(Comparator.reverseOrder()).limit(size).collect(Collectors.toList());
```
Лучше найти популярные фильмы с помощью sql запроса

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/model/Mpa.java:
```diff
@@ -0,0 +1,37 @@
+
+    public Mpa(int id) {
+        this.id = id;
+        switch (id) {
```
Вот тут не совсем понял зачем этот мапинг, в Mpa хранится имя и id в базе данных отдельными полями, их мапинг для получения данных не нужен

slavrikk commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,44 @@
+CREATE TABLE IF NOT EXISTS mpa_rating (
+	mpa INTEGER PRIMARY KEY,
```
Лучше назвать колонку как `id` или `mpa_id`

slavrikk commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,44 @@
+	description VARCHAR(200) NOT NULL,
+	release_date DATE NOT NULL,
+	duration INTEGER NOT NULL,
+	mpa INTEGER REFERENCES mpa_rating (mpa)
```
Лучше назвать колонку как `mpa_id`

DmitryBrosalin committed:
Sprint 12 ver. 12
- Изменил колонку mpa на mpa_id в таблицах и запросах.
- Добавил запрос на получение популярных фильмов через базу данных.

DmitryBrosalin committed:
Sprint 12 ver. 3
- Изменил колонку mpa на mpa_id в таблицах и запросах.
- Добавил запрос на получение популярных фильмов через базу данных.

DmitryBrosalin committed:
Sprint 12 ver. 4
- Изменил логику проверки рейтинга и жанров при сохранении и обновлении фильма, используя базу данных.
